sdk/opa.go

```diff
@@ -247,55 +300,80 @@ func (opa *OPA) Decision(ctx context.Context, options DecisionOptions) (*Decisio
-type DecisionOptions struct {
-	Now   time.Time   // specifies wallclock time used for time.now_ns(), decision log timestamp, etc.
-	Path  string      // specifies name of policy decision to evaluate (e.g., example/allow)
-	Input interface{} // specifies value of the input document to evaluate policy with
-}
+// Partial returns a named decision. This function is threadsafe.
+func (opa *OPA) Partial(ctx context.Context, options PartialOptions) (*DecisionResult, error) {
 
-// DecisionResult contains the output of query evaluation.
-type DecisionResult struct {
-	ID     string      // provides a globally unique identifier for this decision (which is included in the decision log.)
-	Result interface{} // provides the output of query evaluation.
-}
+	if options.Mapper == nil {
+		return nil, fmt.Errorf("PartialQueryMapper must be specified")
+	}
 
-func newDecisionResult() (*DecisionResult, error) {
-	id, err := uuid.New(rand.Reader)
+	record := server.Info{
+		Timestamp: options.Now,
+		Input:     &options.Input,
+		Query:     options.Query,
+		Path:      options.Path,
+	}
+
+	result, err := opa.executeTransaction(
+		ctx,
+		&record,
+		func(result *DecisionResult) {
+			var pq *rego.PartialQueries
+			pq, record.InputAST, record.Bundles, record.Error = partial(ctx, partialEvalArgs{
+				runtime:   opa.state.manager.Info,
+				printHook: opa.state.manager.PrintHook(),
+				compiler:  opa.state.manager.GetCompiler(),
+				store:     opa.state.manager.Store,
+				txn:       record.Txn,
+				now:       record.Timestamp,
+				query:     record.Query,
+				unknowns:  options.Unknowns,
+				input:     *record.Input,
+				m:         record.Metrics,
+			})
+			if record.Error == nil {
+				var recordResult interface{}
+				result.Result, recordResult, record.Error = options.Mapper.MapResults(pq)
+				if record.Error == nil {
+					record.Results = &recordResult
+				}
+			}
+		},
+	)
 	if err != nil {
 		return nil, err
 	}
-	result := &DecisionResult{ID: id}
-	return result, nil
+
+	return result, record.Error
+}
+
+// The first interface being returned is the type that will be used for further processing
+// The second interface is for the decision logger, most of the time these should be the same
+type PartialQueryMapper interface {
+	MapResults(pq *rego.PartialQueries) (interface{}, interface{}, error)
+}
+
+// PartialOptions contains parameters for partial query evaluation.
+type PartialOptions struct {
+	Now      time.Time          // specifies wallclock time used for time.now_ns(), decision log timestamp, etc.
+	Path     string             // specifies name of policy decision to evaluate (e.g., example/allow)
+	Input    interface{}        // specifies value of the input document to evaluate policy with
+	Query    string             // specifies the query to be partially evaluated
+	Unknowns []string           // specifies the unknown elements of the policy
+	Mapper   PartialQueryMapper // specifies the resolver to use when processing results
```

Comment:
nit: the word 'resolver' is leftover here. `s/resolver/mapper/`?

Reply:
looks like I got them all now